### PR TITLE
distinct message for s3 error 'no key' for cases disk and storage

### LINKED ISSUE #54167: make error message with 'no such key' for disk
For now CH writes error log message with 'no such key'. That message can't be used for alerting. Because it could be triggered just by wrong connection string in s3 table function.

We need dedicated message when such error happens with configured s3 disk in tables.

### PR DESCRIPTION
**Pull request by CheSema**
fixes https://github.com/ClickHouse/ClickHouse/issues/54167

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
distinct messages for s3 error 'no key' for cases disk and storage

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit d9b2bc1658fd8f5f93f1d4584f0ccbed7e15d4e0 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61108/d9b2bc1658fd8f5f93f1d4584f0ccbed7e15d4e0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by CheSema** [comment]
need a tests here

File: src/IO/S3/Client.cpp [hunk 1]
```diff
@@ -676,6 +678,23 @@ Client::doRequestWithRetryNetworkErrors(RequestType & request, RequestFn request
...
 }
 
+template <typename RequestResult>
+RequestResult Client::enrichErrorMessage(RequestResult && outcome) const
+{
+    if (outcome.IsSuccess() || !isClientForDisk())
+        return std::move(outcome);
+
+    String enriched_message = fmt::format(
+        "{} {}",
```
Comment (kssenii) [inline src/IO/S3/Client.cpp:688]: Should there be a dot in between? Or it is already in outcome message?

Comment (kssenii) [reply src/IO/S3/Client.cpp:688]: Ah, I see in the test that it is a part of outcome message.

**Review by kssenii — APPROVED** [review]

**Comment by kssenii** [comment]
> Azure does not have separation for metrics counters disk or just function.

https://github.com/ClickHouse/ClickHouse/pull/62458/files

**Comment by CheSema** [comment]
> > Azure does not have separation for metrics counters disk or just function.
> 
> https://github.com/ClickHouse/ClickHouse/pull/62458/files

Yes, I saw it. And only `DiskAzureGetObject` are counted accordingly. But The signal `disk or storage` goes through read setting. Which I removed in that PR.
So I have 3 options:
- support that distinguish flag in Asure client and implement `isClientForDisk` there. That is quite hard. We use client from contribs and that client is marked as `final`.
- revert my changes: remove method `isClientForDisk` from S3 client because there no such method for Asure client. Such information has to be passed through read/write settings. I did not like that way back then and do not like it now. There are places when that settings are default constructed without actual information. I see that info as a property of entire client, not just a parameter of operation.
- and just do not distinguish `DiskAzureGetObject` as we do not distinguish other `DiskAzure*Object` metrics for now. It is the fastest way.

What do you think?

**Comment by kssenii** [comment]
> and just do not distinguish DiskAzureGetObject as we do not distinguish other DiskAzure*Object metrics for now. It is the fastest way.

But we will want to distinguish all metrics and I think it will be required soon, why not support it then?

**Comment by CheSema** [comment]
That changes I what to have in azure client to finish this PR
https://github.com/ClickHouse/azure-sdk-for-cpp/pull/20

**Comment by CheSema** [comment]
> > and just do not distinguish DiskAzureGetObject as we do not distinguish other DiskAzure*Object metrics for now. It is the fastest way.
> 
> But we will want to distinguish all metrics and I think it will be required soon, why not support it then?

I fixed existed azure counters.

**Comment by CheSema** [comment]
There are some tests failed. However I do not see any related tests. So, I wont hold that PR any longer and will merge it.

**Review by kssenii — APPROVED** [review]

**Pull request merged**